txtorcon/socks.py

```diff
@@ -579,7 +581,90 @@ def buildProtocol(self, addr):
 
 
 class SocksError(Exception):
-    pass
+
+    def __init__(self, message, errno=None):
+        super(SocksError, self).__init__(message)
+        self.errno = errno
+
+
+class SucceededReply(SocksError):
+
+    def __init__(self):
+        super(SucceededReply, self).__init__(
+            message='succeeded',
+            errno=0x00)
+
+
+class GeneralServerFailureError(SocksError):
+
+    def __init__(self):
+        super(GeneralServerFailureError, self).__init__(
+            message='general SOCKS server failure',
+            errno=0x01)
+
+
+class ConnectionNotAllowedError(SocksError):
+
+    def __init__(self):
+        super(ConnectionNotAllowedError, self).__init__(
+            message='connection not allowed by ruleset',
+            errno=0x02)
+
+
+class NetworkUnreachableError(SocksError):
+
+    def __init__(self):
+        super(NetworkUnreachableError, self).__init__(
+            message='Network unreachable',
+            errno=0x03)
+
+
+class HostUnreachableError(SocksError):
+
+    def __init__(self):
+        super(HostUnreachableError, self).__init__(
+            message='Host unreachable',
+            errno=0x04)
+
+
+class ConnectionRefusedError(SocksError):
+
+    def __init__(self):
+        super(ConnectionRefusedError, self).__init__(
+            message='Connection refused',
+            errno=0x05)
+
+
+class TtlExpiredError(SocksError):
+
+    def __init__(self):
+        super(TtlExpiredError, self).__init__(
+            message='TTL expired',
+            errno=0x06)
+
+
+class CommandNotSupportedError(SocksError):
+
+    def __init__(self):
+        super(CommandNotSupportedError, self).__init__(
+            message='Command not supported',
+            errno=0x07)
+
+
+class AddressTypeNotSupportedError(SocksError):
+
+    def __init__(self):
+        super(AddressTypeNotSupportedError, self).__init__(
+            message='Address type not supported',
+            errno=0x08)
+
+
+class SocksErrorFactory(object):
+    socks_errors = {cls().errno: cls for cls in SocksError.__subclasses__()}
+
+    @classmethod
+    def create(cls, errno):
+        return cls.socks_errors[errno]()
 
```

Comment:
Cool, I completely agree. Now that I'm thinking again, `errno` was also a bad choice because we are dealing with reply codes actually. Thanks!